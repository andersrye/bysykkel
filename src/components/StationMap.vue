<template>
  <div ref="mapContainer" />
</template>

<script setup>
import { computed, reactive, watch, ref, watchEffect } from "vue"
import mapboxgl from "mapbox-gl"
import StationPopupContent from '@/components/StationPopupContent'
import "mapbox-gl/dist/mapbox-gl.css"
import { useMapbox } from "@/composables/mapbox"
import { stationsAsFeatureCollection } from '@/utils/geojson'
import circleIcon from '@/assets/circle_24.png'
import availableBikesLayer from '@/map-layers/available-bikes.json'
import availableDocksLayer from '@/map-layers/available-docks.json'
import stationLayer from '@/map-layers/stations.json'

const mapContainer = ref(null)
const reactiveAvailableBikesLayer = reactive(availableBikesLayer)
const reactiveAvailableDocksLayer = reactive(availableDocksLayer)
const geoJson = computed(() => stationsAsFeatureCollection(props.stations))

const props = defineProps({
  stations: { type: Array, default: () => ([]) },
  selectedStation: { type: Object, default: null },
  showBikes: {type: Boolean, default: true}
})

const map = useMapbox({
  container: mapContainer,
  accessToken: 'pk.eyJ1IjoiYW5kZXJzcnllIiwiYSI6ImNsMHI1emE5aDAwNGMzaW5tOHZrdmt0azcifQ.VCTJiGvc-mef33wXhlqk7g',
  style: "mapbox://styles/mapbox/streets-v11",
  zoom: 5,
  center: [10.735834, 60.917947]
})

map.addImage('circle-icon', circleIcon, {sdf: true})

map.addGeoJsonSource('stations', geoJson)

map.addLayer(stationLayer)
map.addLayer(reactiveAvailableBikesLayer)
map.addLayer(reactiveAvailableDocksLayer)

map.on('mouseenter', [availableBikesLayer.id, availableDocksLayer.id], () => {
  map.mapboxMap.value.getCanvas().style.cursor = 'pointer';
})

map.on('mouseleave', [availableBikesLayer.id, availableDocksLayer.id], () => {
  map.mapboxMap.value.getCanvas().style.cursor = '';
})

map.on('click', [availableBikesLayer.id, availableDocksLayer.id], (e) => {
  const feature = e.features[0]
  if(feature) {
    openPopup(feature.properties, e.lngLat)
  }
})

function toggleLayers(showBikes) {
  console.log('toggleLayers', showBikes)
  reactiveAvailableBikesLayer.layout.visibility = showBikes ? 'visible' : 'none'
  reactiveAvailableDocksLayer.layout.visibility = showBikes ? 'none' : 'visible'
}

function openPopup(station, lngLat) {
  map.openPopup(StationPopupContent, lngLat, {station})
}

watchEffect(() => toggleLayers(props.showBikes))

watch(() => props.selectedStation, station => {
  if (station) {
    const lngLat = [station.lon, station.lat]
    map.mapboxMap.value?.easeTo({
      center: lngLat,
      zoom: 16
    })
    openPopup(station, lngLat)
  }
})

//fit map to geoJson once, then remove watch
const unwatch = watchEffect(() => {
  if (map.mapboxMap.value && geoJson.value?.features?.length) {
    const coordinates = geoJson.value.features.map(f => f.geometry.coordinates)
    const bounds = new mapboxgl.LngLatBounds(coordinates[0], coordinates[0])
    for (const coordinate of coordinates) {
      bounds.extend(coordinate)
    }
    map.mapboxMap.value?.fitBounds(bounds, {
      padding: { top: 50, bottom: 50, left: 50, right: 50 },
      duration: 1000,
    })
    unwatch()
  }
})

</script>
