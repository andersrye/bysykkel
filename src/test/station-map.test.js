import { mount } from '@vue/test-utils'
import StationMap from '@/components/StationMap.vue'
import { expect, test, vi } from 'vitest'
import mapboxgl from "mapbox-gl";
import station from './test-data/station-combined.json'

//mapbox can't really work in jsdom, so we have to mock it
vi.mock("mapbox-gl", () => {
    const Map = vi.fn()
    Object.assign(Map.prototype, {
        once: vi.fn((type, callback) => {
            if(type === 'load') callback() //call the callback to trigger everything that depends on the map being loaded
        }),
        on: vi.fn(),
        off: vi.fn(),
        loadImage: vi.fn(),
        getSource: vi.fn(),
        addSource: vi.fn(),
        getLayer: vi.fn(),
        addLayer: vi.fn(),
        easeTo: vi.fn()
    })
    const Popup = vi.fn()
    Object.assign(Popup.prototype, {
        remove: vi.fn(),
        addTo: vi.fn(),
        setHTML: vi.fn().mockReturnThis(), //in order to be chainable these have to return ´this´
        setLngLat: vi.fn().mockReturnThis(),
        on: vi.fn().mockReturnThis()
    })
    return {
        default: {
            Map,
            Popup
        }
    }
})

describe('app', () => {

    afterEach(() => {
        vi.clearAllMocks()
    })

    test('mount component', async () => {
        const wrapper = mount(StationMap)
        expect(wrapper.html()).toMatchSnapshot()
    })

    test('selectedStation should pan the map and open a popup', async () => {
        const wrapper = mount(StationMap)
        await wrapper.setProps({selectedStation: station})

        const lngLat = [station.lon, station.lat]
        const mapboxInstance = mapboxgl.Map.mock.instances[0]
        const popupInstance = mapboxgl.Popup.mock.instances[0]

        expect(mapboxInstance.easeTo).toHaveBeenCalledWith(expect.objectContaining({center: lngLat}))
        expect(popupInstance.setLngLat).toHaveBeenCalledWith(lngLat)
        expect(popupInstance.addTo).toHaveBeenCalled()

        //todo: check that PopupContent has been mounted properly?
    })
})
