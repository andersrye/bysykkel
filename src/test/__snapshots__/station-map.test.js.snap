// Vitest Snapshot v1

exports[`StationMap > mount component 1`] = `"<div></div>"`;
